common/cpp/src/google_smart_card_common/value_nacl_pp_var_conversion.cc

```diff
@@ -100,4 +154,37 @@ pp::Var ConvertValueToPpVar(const Value& value) {
   GOOGLE_SMART_CARD_NOTREACHED;
 }
 
+optional<Value> ConvertPpVarToValue(const pp::Var& var,
+                                    std::string* error_message) {
+  if (var.is_undefined()) return Value();
+  if (var.is_null()) return Value();
+  if (var.is_bool()) return Value(var.AsBool());
+  if (var.is_string()) return Value(var.AsString());
+  if (var.is_object()) {
+    if (error_message) {
+      *error_message =
+          FormatPrintfTemplate(kUnsupportedPpVarTypeError, "object");
+    }
+    return {};
+  }
+  if (var.is_array())
+    return CreateValueFromPpVarArray(pp::VarArray(var), error_message);
+  if (var.is_dictionary()) {
+    return CreateValueFromPpVarDictionary(pp::VarDictionary(var),
+                                          error_message);
+  }
+  if (var.is_resource()) {
+    if (error_message) {
+      *error_message =
+          FormatPrintfTemplate(kUnsupportedPpVarTypeError, "resource");
+    }
+    return {};
+  }
+  if (var.is_int()) return Value(var.AsInt());
+  if (var.is_double()) return Value(var.AsDouble());
```

Comment:
Open question: Should we actually return an integer Value in case the passe double value is actually an integer and fits within the range of precisely representable integer numbers (which is typically 53 bits)?

I do this in the existing code that, for example, converts between `pp::Var` and `int64_t`, but in that code we always know the target type. Here in this PR we don't know how the resulting `Value` will be used (as a double or as an integer), so it's up to us whether we want to try attempting the casting early...

Reply:
Good question.

I think it depends on the user expectation of such functionality as Value::Type would be different in both cases.

IMHO, I am leaning towards keeping it as it is (return Value::Type::kFloat value) and let the user decide what they want to do with the value.

The only use-case I find it beneficial to return integer is: since pp:Var only supports int32_t, the user will have to use double instead, in this case we give the user a ready-to-use Value::Type. (Instead they would need to cast).
But at the same time, we would hurt other use-cases where the user really expects Value::Type::kFloat (even if the value happens to be an integer).

Reply:
I also think we should return a double here - it would be confusing if one converts a double and in some cases it comes out as int.

Reply:
Thanks, Mahmoud and Fabian. That makes sense, I'll keep that behavior in the PR.